lib/eth/tx.rb

```diff
@@ -70,19 +78,28 @@ def to_h
     end
 
     def from
-      if signature
-        public_key = OpenSsl.recover_compact(signature_hash, signature)
+      if ecdsa_signature
+        public_key = OpenSsl.recover_compact(signature_hash, ecdsa_signature)
         Utils.public_key_to_address(public_key) if public_key
       end
     end
 
     def signature
       return @signature if @signature
-      self.signature = [
-        Utils.int_to_base256(v),
-        Utils.zpad_int(r),
-        Utils.zpad_int(s),
-      ].join if [v, r, s].all?
+      @signature = { v: v, r: r, s: s } if [v, r, s].all? && (v > 0)
+    end
+
+    def ecdsa_signature
+      return @ecdsa_signature if @ecdsa_signature
+      
+      if [v, r, s].all? && (v > 0)
+        s_v = (self.chain_id) ? (v - (self.chain_id * 2) - 8) : v
+        @ecdsa_signature = [
```

Comment:
Argh I just noticed you comment about Key.personal_recover. I'll take a look and get back to you on this.

Reply:
I looked at the behavior of Key.personal_recover, and also at personal.recover in geth (I'm at 1.8.14). personal.recover only accepts ECDSA signatures, where V is either 27 or 28. Since Key.sign on this branch only places 27 or 28 in V, the generated personal signature is valid.

As a matter of fact, the master branch is buggy: if you set the chain ID to other than nil in Eth, the generated personal signature contains the EIP-155 adjustment and is invalid. So this pull requests fixes both bugs.